doc/decisions/0_drafts/key_value.md

```diff
@@ -0,0 +1,123 @@
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+### Read-only values
+
+A simple way to define an API would be to make everything read-only.
+A `Key` can be created with a certain value and that's it.
+If you want a different value, create a new `Key`.
+
+That would certainly simplify some parts of the implementation, but it would be major inconvenience for the caller.
+Also allocating entirely new `Key`s just to change values, could lead to unnecessary, short-lived heap-allocations.
+
+### Direct access
+
+The next simplest way to define the API, is to give the caller full direct access.
+
+The caller creates a `Key` with a `void *` and a `size_t`.
+Afterwards the caller an access and change them at any time in any way they want to.
+
+Basically, a `Key` would not have ownership of its value.
+It would just have a reference to it.
+
+This solution could easily lead to various lifetime problems.
+By not owning the value, the `Key` cannot be reponsible for freeing it.
+
+### Buffers provided by caller
+
+To fix the lifetime problems from above, a `Key` should own its value.
+
+That means when a `Key` is created with a `void *` and a `size_t`, `libelektra-core` does a `memcpy` to create a copy of the data.
+Changing the value works the same way.
+
+When the caller wants access to that the value again, they must provide a suitably sized buffer as a `void *` and `libelektra-core` will do a `memcpy` to give a copy of the value to the caller.
+
+This solves all lifetime problems, because `Key` fully owns its value and never even shares a reference with anyone.
+However, the `memcpy` on access is totally unnecessary when we just want to read the value.
+
+### Semi-direct access
+
+To avoid the unnecessary `memcpy` from above, the API should be asymmetric.
+A `Key` owns its value, but it also provides read-only references to the caller.
+
+So when creating a `Key` or changing its value, `libelektra-core` does a `memcpy`.
+But when the caller wants access to the value, `libelektra-core` simply returns a `const` pointer.
+The caller can read the value, but cannot change it.
+If they want a copy they can modify, the caller must do a `memcpy` themselves.
+
+However, by omitting the `memcpy` for reading values, the value inside a `Key` is no longer immutable.
+Take this example:
+
+```c
+typedef struct {
+    int a;
+    char * b;
+} foo;
+
+int main (void) {
+    char arr[] = "test";
+    foo f = {.a = 1, .b = arr };
+
+    foo g;
+    memcpy(&g, &f, sizeof(foo));
+
+    const foo * h = &g;
+
+    // h->a = 7;   // ERROR: assignment of member 'a' in read-only object
+    h->b[0] = 'b'; // works fine, b is not const and C doesn't propagte the const from h to members
+
+    return 0;
+}
+```
+
+This may become a problem, when values are shared between `Key`s via [Copy on Write](../2_in_progress/copy_on_write.md) copies.
+The only solution here is to clearly document, that value returned from a `Key` must not be modified in any way.
```

Comment:
1. `mprotect` works on a page level and is POSIX-only. It's therefore out of the question.
2. I don't think returning a `const` pointer is enough documentation. It's not uncommon to have a C API that returns a const pointer, but it's safe to cast it to non-const, e.g. if the pointer is passed through from the caller (C doesn't have a way to preserve constness in that way). But even if we ignore those kinds of API, we should at least document **why** it is unsafe to modify the value (even if allowed by the compiler). But the file just says "clearly document" for a reason, we don't need to decide what exactly this means here.
3. I don't think implications like this should go unmentioned just because they are status quo. The status quo is known now, but if somebody reads this decision in the future, the status quo will be what we decide here. So they might assume we did not consider this issue and try to reopen the decision because of it. But we did in fact consider the issue and explicitly decided it is acceptable considering the alternatives.